### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by rafaelrpbelo**
In this PR the follow steps are added:
- Add Capybara's gem
- Create Customer's CRUD
- Create Vehicle's CRUD
- Set locale to pt-BR

### CODE REVIEW

File: db/migrate/20151113223049_create_customer.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+      t.string :postal_code, limit: 9
+      t.string :address
+      t.string :neighborhood
+      t.string :adjunct
+      t.string :city
+      t.string :state
+      t.string :phone
+      t.string :cell_phone
+      t.string :email
+    end
```
Comment (dmitryrck) [inline db/migrate/20151113223049_create_customer.rb:26]: Missing `t.timestamps` :P

File: app/controllers/customers_controller.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  end
+
+  def create
+    @customer = Customer.new(customer_params)
+
+    if @customer.save
+      flash[:notice] = 'Customer created successful'
+      redirect_to @customer
+    else
+      flash[:alert] = "Customer can't be created"
```
Comment (dmitryrck) [inline app/controllers/customers_controller.rb]: I don't understand: why did you change locale to _pt-br_, but you are using _en_ here?

File: app/models/customer.rb [hunk 3]
```diff
@@ -0,0 +1,2 @@
+class Customer < ActiveRecord::Base
+end
```
Comment (dmitryrck) [inline app/models/customer.rb:21]: it is missing validation, they are important :|

File: app/models/vehicle.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
+class Vehicle < ActiveRecord::Base
+  STATUS = {
+    :stock => I18n.t('stock', scope: 'enumerize.vehicle.status'),
+    :sold => I18n.t('sold', scope: 'enumerize.vehicle.status')
+  }
+
+  def status
+    I18n.t(read_attribute(:status), scope: 'enumerize.vehicle.status').capitalize
+  end
```
Comment (dmitryrck) [inline app/models/vehicle.rb:18]: I like the gem [enumerate_it](https://github.com/cassiomarques/enumerate_it).

File: app/views/customers/_form.html.erb [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  <%= f.text_field :phone %>
+
+  <%= f.label :cell_phone %>
+  <%= f.text_field :cell_phone %>
+
+  <%= f.label :email %>
+  <%= f.text_field :email %>
+
+  <%= f.submit %>
+<% end %>
```
Comment (dmitryrck) [inline app/views/customers/_form.html.erb]: For a app that it has too many forms or forms can be too big I enjoy simple_form gem.

File: app/views/customers/show.html.erb [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+<%= @customer.father %>
+<%= @customer.postal_code %>
+<%= @customer.address %>
+<%= @customer.neighborhood %>
+<%= @customer.adjunct %>
+<%= @customer.city %>
+<%= @customer.state %>
+<%= @customer.phone %>
+<%= @customer.cell_phone %>
+<%= @customer.email %>
```
Comment (dmitryrck) [inline app/views/customers/show.html.erb:20]: There is a gem called show_for, I enjoy it to use in that some case of simple_form.

File: spec/features/customer_spec.rb [hunk 7]
```diff
@@ -0,0 +1,105 @@
+require 'rails_helper'
+
+describe Customer do
+  it 'create' do
+    visit '/customers/new'
```
Comment (dmitryrck) [inline spec/features/customer_spec.rb]: `visit new_customer_path`

File: spec/features/customer_spec.rb [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+      neighborhood: 'Ininga',
+      adjunct: '',
+      city: 'Teresina',
+      state: 'PI',
+      phone: '(86) 3232-3201',
+      cell_phone: '(86) 99900-1213',
+      email: 'customer@test.com'
+    )
+
+    visit edit_customer_path(customer)
```
Comment (dmitryrck) [inline spec/features/customer_spec.rb]: Use Rspec like a user:

```
visit root_path # or: _visit customers_path_ if you had not set root_path
click_on "Customer Test" # or: _click_on "Edit"_
```

Your users will get an id of a customer and build the url by hand? I don't think so, so: use rspec here like a end user.

File: spec/models/customer_spec.rb [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+      phone: '(86) 3232-3201',
+      cell_phone: '(86) 99900-1213',
+      email: 'customer@test.com'
+    )
+  end
+
+  it { should be_valid }
+
+  it 'should can create a new user' do
+    subject.save
```
Comment (dmitryrck) [inline spec/models/customer_spec.rb]: I prefer to use a block:

```
expect { subject.save }.to change { subject.persisted? }
```

**Comment by dmitryrck** [comment]
Well… if you know your app will grow use gems like: `simple_form`, `show_for`, `enumerate_it`, `responders`, …

**Comment by dmitryrck** [comment]
Don't use locale until all your views is in the target locale :/

So… You could split this two model creation into two commits :P

File: db/migrate/20151113223049_create_customer.rb (hunk 1 above)
Comment (rafaelrpbelo) [inline db/migrate/20151113223049_create_customer.rb:26]: :dizzy_face:

File: app/controllers/customers_controller.rb (hunk 2 above)
Comment (rafaelrpbelo) [inline app/controllers/customers_controller.rb]: Because I put i18n before. :dizzy_face:

File: app/models/customer.rb (hunk 3 above)
Comment (rafaelrpbelo) [inline app/models/customer.rb:21]: Yes, I put off for later and I forgot :neutral_face:

File: app/models/vehicle.rb (hunk 4 above)
Comment (rafaelrpbelo) [inline app/models/vehicle.rb:18]: Commonly I use `enumerize`, but how I use this just here I not change yet.
I'm walking like baby steps. :baby:

File: app/views/customers/_form.html.erb (hunk 5 above)
Comment (rafaelrpbelo) [inline app/views/customers/_form.html.erb]: Yes. I think this already the time to use! Thx man ;)

File: app/views/customers/show.html.erb (hunk 6 above)
Comment (rafaelrpbelo) [inline app/views/customers/show.html.erb:20]: For while I'm just doing the basic presentation, but I like the tip! I didn't know this gem. :wink:

File: spec/features/customer_spec.rb (hunk 7 above)
Comment (rafaelrpbelo) [inline spec/features/customer_spec.rb]: I forgot this too! :disappointed_relieved:

File: spec/features/customer_spec.rb (hunk 8 above)
Comment (rafaelrpbelo) [inline spec/features/customer_spec.rb]: Yes sir! Nice observation! :grin:

File: spec/models/customer_spec.rb (hunk 9 above)
Comment (rafaelrpbelo) [inline spec/models/customer_spec.rb]: Nice! :grin:

**Comment by rafaelrpbelo** [comment]
@dmitryrck review again plz, bro!

File: app/models/customer.rb [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+  validates_format_of :email, with: /\A([^@\s]+)@((?:[-a-z0-9]+\.)+[a-z]{2,})\z/i
+  validate :old_enough
+  validate :rg_expedition_date
+
+  private
+
+  # 18 years old or more
+  def old_enough
+    if self.birth_date
+      old_insufficient = (self.birth_date + 18.years) > Date.today
```
Comment (dmitryrck) [inline app/models/customer.rb]: Humm… 

```
if self.birth_date.present? && self.birth_date > 18.years.ago
  errors.add(:birth_date)
end
```

?

File: app/models/customer.rb [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+      errors.add(:birth_date) if old_insufficient
+    end
+  end
+
+  def rg_expedition_date
+    errors.add(:rg_expedition) if self.rg_expedition && self.rg_expedition > Date.today
```
Comment (dmitryrck) [inline app/models/customer.rb]: Use `Date.current`.

File: spec/models/customer_spec.rb [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+  it { is_expected.to validate_length_of(:place_birth_uf).is_equal_to(2) }
+
+  it 'should require email has a valid format' do
+    subject.email = 'jhon.doe@example.com'
+    expect(subject).to be_valid
+  end
+
+  context 'date' do
+    it 'should require old should to be enough (18 or more)' do
+      expect{ subject.birth_date = Date.today - 17.years }.to change { subject.valid? }.to false
```
Comment (dmitryrck) [inline spec/models/customer_spec.rb]: `Date.current`

Comment (dmitryrck) [inline spec/models/customer_spec.rb]:
```
expect { subject.birth_date = 17.years.ago }.to change { subject.valid? }.to false
```

But I prefer in this way:

```
  subject.birth_date = 17.years.ago
  expect(subject).not_to be_valid
```

File: spec/models/customer_spec.rb [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+      expect(subject.errors.full_messages).to include 'Data de nascimento não é válido'
+    end
+
+    it 'should require rg_expedition should be equal or less the current date' do
+      expect{ subject.rg_expedition = Date.today + 1 }.to change { subject.valid? }.to false
```
Comment (dmitryrck) [inline spec/models/customer_spec.rb]: `Date.today + 1` => `Date.tomorrow`

File: db/migrate/20151113223049_create_customer.rb [hunk 14]
```diff
@@ -0,0 +1,28 @@
+class CreateCustomer < ActiveRecord::Migration
+  def change
+    create_table :customers do |t|
+      t.string :full_name, null: false
+      t.string :cpf, null: false
```
Comment (dmitryrck) [inline db/migrate/20151113223049_create_customer.rb]: You want to add `uniqueness` validation to CPF :|

File: db/migrate/20151113223049_create_customer.rb [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+      t.string :father
+      t.string :postal_code, limit: 9
+      t.string :address
+      t.string :neighborhood
+      t.string :adjunct
+      t.string :city
+      t.string :state
+      t.string :phone
+      t.string :cell_phone
+      t.string :email
```
Comment (dmitryrck) [inline db/migrate/20151113223049_create_customer.rb]: To the e-mail field it is a good idea too.

File: db/migrate/20151109222454_create_vehicle.rb [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+      t.string :license_plate, limit: 7, null: false
+      t.string :city
+      t.string :uf, limit: 2
+      t.string :color
+      t.string :fuel
+      t.string :renavam
+      t.string :chassi
+      t.string :owner
+      t.string :old_license_plate
+      t.string :status
```
Comment (dmitryrck) [inline db/migrate/20151109222454_create_vehicle.rb:17]: It is also a good idea to add uniqueness validation to `renavam` and `chassi`, is it not?

**Comment by dmitryrck** [comment]
√ review again

File: app/models/customer.rb (hunk 10 above)
Comment (rafaelrpbelo) [inline app/models/customer.rb]: What's wrong with this?

File: spec/models/customer_spec.rb (hunk 12 above)
Comment (rafaelrpbelo) [inline spec/models/customer_spec.rb]: yes.. look good for me!

File: spec/models/customer_spec.rb (hunk 13 above)
Comment (rafaelrpbelo) [inline spec/models/customer_spec.rb]: nuss.. I really forgot this!

File: app/models/customer.rb (hunk 10 above)
Comment (dmitryrck) [inline app/models/customer.rb]: Your way is not good to read:

```
if condition
  action if other_condition
end
```

It is more clean when you write/read:

```
if condition
  action
end
```

Comment (rafaelrpbelo) [inline app/models/customer.rb]: Done: _ca7843d_

File: app/models/customer.rb (hunk 11 above)
Comment (rafaelrpbelo) [inline app/models/customer.rb]: Done: _ca7843d_

File: spec/models/customer_spec.rb (hunk 12 above)
Comment (rafaelrpbelo) [inline spec/models/customer_spec.rb]: Done: _ca7843d_

File: spec/models/customer_spec.rb (hunk 13 above)
Comment (rafaelrpbelo) [inline spec/models/customer_spec.rb]: Done: _ca7843d_

File: db/migrate/20151113223049_create_customer.rb (hunk 14 above)
Comment (rafaelrpbelo) [inline db/migrate/20151113223049_create_customer.rb]: Done:  https://github.com/rafaelrpbelo/car_dealership_sm/commit/28624af6b42bd6f929ab6f8d3f7891020f42212c

File: db/migrate/20151109222454_create_vehicle.rb (hunk 16 above)
Comment (rafaelrpbelo) [inline db/migrate/20151109222454_create_vehicle.rb:17]: Done:  https://github.com/rafaelrpbelo/car_dealership_sm/commit/28624af6b42bd6f929ab6f8d3f7891020f42212c

File: db/migrate/20151113223049_create_customer.rb (hunk 15 above)
Comment (rafaelrpbelo) [inline db/migrate/20151113223049_create_customer.rb]: Done:  https://github.com/rafaelrpbelo/car_dealership_sm/commit/28624af6b42bd6f929ab6f8d3f7891020f42212c

**Comment by rafaelrpbelo** [comment]
@dmitryrck If alright, you can to merge plz?

File: app/helpers/bootstrap_helper.rb [hunk 17]
```diff
@@ -0,0 +1,12 @@
+module BootstrapHelper
+  # Navigation
+  def active_link_to(name, link, params={})
+    content_tag :li, class: active?(link) do
+      link_to name, link, class: params[:class]
+    end
+  end
+
+  def active?(link)
```
Comment (dmitryrck) [inline app/helpers/bootstrap_helper.rb:11]: This method could be private :P

File: app/views/customers/index.html.erb [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+
+<table>
+  <thread>
+    <tr>
+      <th>Nome</th>
+      <th></th>
+    </tr>
+  </thread>
+  <tbody>
+    <%= content_tag_for :tr, @customers do |customer| %>
```
Comment (dmitryrck) [inline app/views/customers/index.html.erb:11]: Good to know.

File: app/helpers/bootstrap_helper.rb (hunk 17 above)
Comment (rafaelrpbelo) [inline app/helpers/bootstrap_helper.rb:11]: Done: https://github.com/rafaelrpbelo/car_dealership_sm/commit/d6dc24c259408f906eda496aaecf9c1a2bc7163d
